File: typescript/events/emitter.ts
import type { SpiderEvents, SpiderEventName } from './types.js';

type Handler<T> = (data: T) => void;

/** Type-safe event emitter for SpiderBrowser events. */
export class SpiderEventEmitter {
  private handlers = new Map<string, Set<Handler<any>>>();

  on<K extends SpiderEventName>(event: K, handler: Handler<SpiderEvents[K]>): this {
    let set = this.handlers.get(event);
    if (!set) {
      set = new Set();
      this.handlers.set(event, set);
    }
    set.add(handler);
    return this;
  }

  off<K extends SpiderEventName>(event: K, handler: Handler<SpiderEvents[K]>): this {
    this.handlers.get(event)?.delete(handler);
    return this;
  }

  once<K extends SpiderEventName>(event: K, handler: Handler<SpiderEvents[K]>): this {
    const wrapped: Handler<SpiderEvents[K]> = (data) => {
      this.off(event, wrapped);
      handler(data);
    };
    return this.on(event, wrapped);
  }

  emit<K extends SpiderEventName>(event: K, data: SpiderEvents[K]): void {
    const set = this.handlers.get(event);
    if (set) {
      for (const handler of set) {
        try {
          handler(data);
        } catch {
          // Don't let user handlers crash the library
        }
      }
    }
  }

  removeAllListeners(event?: SpiderEventName): void {
    if (event) {
      this.handlers.delete(event);
    } else {
      this.handlers.clear();
    }
  }
}

File: python/spider_browser/events/emitter.py
"""Typed event emitter for SpiderBrowser events."""

from __future__ import annotations
from collections import defaultdict
from typing import Any, Callable


class SpiderEventEmitter:
    """Simple synchronous event emitter."""

    def __init__(self) -> None:
        self._handlers: dict[str, list[Callable[..., Any]]] = defaultdict(list)

    def on(self, event: str, handler: Callable[..., Any]) -> None:
        self._handlers[event].append(handler)

    def off(self, event: str, handler: Callable[..., Any]) -> None:
        try:
            self._handlers[event].remove(handler)
        except ValueError:
            pass

    def once(self, event: str, handler: Callable[..., Any]) -> None:
        def wrapper(*args: Any, **kwargs: Any) -> None:
            self.off(event, wrapper)
            handler(*args, **kwargs)
        self._handlers[event].append(wrapper)

    def emit(self, event: str, data: dict[str, Any] | None = None) -> None:
        for handler in list(self._handlers.get(event, [])):
            try:
                handler(data or {})
            except Exception:
                pass

    def remove_all_listeners(self, event: str | None = None) -> None:
        if event:
            self._handlers.pop(event, None)
        else:
            self._handlers.clear()
